File: C/Queue/Queue_06.c
/*
===============================================================================
Program Name    : Queue Menu Driven Application
Program Number  : 007
File Name       : Queue07.c

Description     : This program demonstrates a complete
                  Menu Driven Queue using Singly Linear
                  Linked List.

Date            : August 2026
Language        : C
Author          : Atharv Tushar Bhosale
===============================================================================
*/

#include<stdio.h>
#include<stdlib.h>

struct node
{
    int data;
    struct node *next;
};

typedef struct node NODE;
typedef struct node* PNODE;
typedef struct node** PPNODE;

///////////////////////////////////////////////////////////////////////////////
//
// Function Name : Enqueue
// Description   : Insert element at Last
//
///////////////////////////////////////////////////////////////////////////////

void Enqueue(PPNODE first,int iNo)
{
    PNODE newn = NULL;
    PNODE temp = NULL;

    newn = (PNODE)malloc(sizeof(NODE));

    newn->data = iNo;
    newn->next = NULL;

    if(*first == NULL)
    {
        *first = newn;
    }
    else
    {
        temp = *first;

        while(temp->next != NULL)
        {
            temp = temp->next;
        }

        temp->next = newn;
    }
}

///////////////////////////////////////////////////////////////////////////////
//
// Function Name : Dequeue
// Description   : Delete element from First
//
///////////////////////////////////////////////////////////////////////////////

int Dequeue(PPNODE first)
{
    PNODE temp = NULL;
    int iValue = 0;

    if(*first == NULL)
    {
        printf("Queue is Empty\n");
        return -1;
    }

    iValue = (*first)->data;

    temp = *first;

    *first = (*first)->next;

    free(temp);

    return iValue;
}

///////////////////////////////////////////////////////////////////////////////
//
// Function Name : Display
// Description   : Display Queue
//
///////////////////////////////////////////////////////////////////////////////

void Display(PNODE first)
{
    printf("\nQueue Elements :\n");

    while(first != NULL)
    {
        printf("|%d| -> ",first->data);
        first = first->next;
    }

    printf("NULL\n");
}

///////////////////////////////////////////////////////////////////////////////
//
// Function Name : Count
// Description   : Count number of nodes
//
///////////////////////////////////////////////////////////////////////////////

int Count(PNODE first)
{
    int iCount = 0;

    while(first != NULL)
    {
        iCount++;
        first = first->next;
    }

    return iCount;
}

///////////////////////////////////////////////////////////////////////////////
//
// Entry Point Function
//
///////////////////////////////////////////////////////////////////////////////

int main()
{
    PNODE head = NULL;

    int iChoice = 0;
    int iValue = 0;
    int iRet = 0;

    printf("-------------------------------------------------\n");
    printf("          Queue Menu Driven Application\n");
    printf("-------------------------------------------------\n");

    while(1)
    {
        printf("\n-------------------------------------------------\n");
        printf("\n---------------------QUEUE-----------------------\n");
        printf("\n-------------------------------------------------\n");
        printf("1. Enqueue\n");
        printf("2. Dequeue\n");
        printf("3. Display\n");
        printf("4. Count\n");
        printf("5. Exit\n");
        printf("-------------------------------------------------\n");

        printf("Enter your choice : ");
        scanf("%d",&iChoice);

        switch(iChoice)
        {
            case 1:

                printf("Enter the data : ");
                scanf("%d",&iValue);

                Enqueue(&head,iValue);

                printf("Element inserted successfully\n");
                break;

            case 2:

                iRet = Dequeue(&head);

                if(iRet != -1)
                {
                    printf("Removed Element : %d\n",iRet);
                }
                break;

            case 3:

                Display(head);
                break;

            case 4:

                iRet = Count(head);

                printf("Number of Elements : %d\n",iRet);
                break;

            case 5:

                printf("Thank you for using Queue Application\n");
                return 0;

            default:

                printf("Invalid Choice\n");
                break;
        }
    }

    return 0;
}

/*
===============================================================================
Sample Output
-------------------------------------------------
---------------------QUEUE-----------------------
-------------------------------------------------
1. Enqueue
2. Dequeue
3. Display
4. Count
5. Exit
-------------------------------------------------

Enter your choice : 1
Enter the data : 11

Enter your choice : 1
Enter the data : 21

Enter your choice : 3

Queue Elements :
|11| -> |21| -> NULL

Enter your choice : 2

Removed Element : 11

Queue Elements :
|21| -> NULL

===============================================================================
End of Program
===============================================================================
*/

File: Java/Queue/Queue_06.java
/*
===============================================================================
Program Name    : Complete Queue Application
Program Number  : 006
File Name       : Queue_06.java

Description     : This program demonstrates all Queue
                  operations using Singly Linear Linked
                  List.

                  Operations:
                  1. Enqueue
                  2. Dequeue
                  3. Display
                  4. Count

Concepts Used:
1. Class and Object
2. Linked List
3. Queue Data Structure
4. Dynamic Memory Allocation

Date            : August 2026
Language        : Java
Author          : Atharv Tushar Bhosale
===============================================================================
*/

class node
{
    public int data;
    public node next;
}

class Queue
{
    private node first;
    private int iCount;

    public Queue()
    {
        first = null;
        iCount = 0;
    }

    ///////////////////////////////////////////////////////////////////////////
    //
    // Function Name : Enqueue
    // Description   : Insert element at Last
    //
    ///////////////////////////////////////////////////////////////////////////

    public void Enqueue(int iNo)
    {
        node newn = null;
        node temp = null;

        newn = new node();

        newn.data = iNo;
        newn.next = null;

        if(first == null)
        {
            first = newn;
        }
        else
        {
            temp = first;

            while(temp.next != null)
            {
                temp = temp.next;
            }

            temp.next = newn;
        }

        iCount++;
    }

    ///////////////////////////////////////////////////////////////////////////
    //
    // Function Name : Dequeue
    // Description   : Delete element from First
    //
    ///////////////////////////////////////////////////////////////////////////

    public int Dequeue()
    {
        int iValue = 0;
        node temp = null;

        if(first == null)
        {
            System.out.println("Queue is Empty");
            return -1;
        }

        iValue = first.data;

        temp = first;
        first = first.next;
        temp = null;

        iCount--;

        return iValue;
    }

    ///////////////////////////////////////////////////////////////////////////
    //
    // Function Name : Display
    // Description   : Display Queue
    //
    ///////////////////////////////////////////////////////////////////////////

    public void Display()
    {
        node temp = first;

        System.out.println("\nElements of Queue are : ");

        while(temp != null)
        {
            System.out.print("|"+temp.data+"| -> ");
            temp = temp.next;
        }

        System.out.println("NULL");
    }

    ///////////////////////////////////////////////////////////////////////////
    //
    // Function Name : Count
    // Description   : Count number of elements
    //
    ///////////////////////////////////////////////////////////////////////////

    public int Count()
    {
        return iCount;
    }
}

class Queue_06
{
    public static void main(String A[])
    {
        Queue qobj = new Queue();

        int iRet = 0;

        qobj.Enqueue(11);
        qobj.Enqueue(21);
        qobj.Enqueue(51);
        qobj.Enqueue(101);

        qobj.Display();

        iRet = qobj.Count();
        System.out.println("\nNumber of Elements : "+iRet);

        iRet = qobj.Dequeue();
        System.out.println("\nRemoved Element : "+iRet);

        qobj.Display();

        iRet = qobj.Count();
        System.out.println("\nNumber of Elements : "+iRet);

        iRet = qobj.Dequeue();
        System.out.println("\nRemoved Element : "+iRet);

        qobj.Display();

        iRet = qobj.Count();
        System.out.println("\nNumber of Elements : "+iRet);
    }
}

/*
===============================================================================
Sample Output

Elements of Queue are :

|11| -> |21| -> |51| -> |101| -> NULL

Number of Elements : 4

Removed Element : 11

Elements of Queue are :

|21| -> |51| -> |101| -> NULL

Number of Elements : 3

Removed Element : 21

Elements of Queue are :

|51| -> |101| -> NULL

Number of Elements : 2

===============================================================================
End of Program
===============================================================================
*/
